consensus-types/hdiff/property_test.go

```diff
@@ -0,0 +1,391 @@
+		source, _ := util.DeterministicGenesisStateElectra(t, validators)
+		target := source.Copy()
+		
+		// Apply realistic changes
+		_ = target.SetSlot(source.Slot() + primitives.Slot(slots))
+		
+		if changes > 0 {
+			validatorList := target.Validators()
+			for i := 0; i < changes && i < len(validatorList); i++ {
+				validatorList[i].EffectiveBalance += 1000000000 // 1 ETH
+			}
+			_ = target.SetValidators(validatorList)
+		}
+		
+		// Operations should complete quickly
+		start := time.Now()
+		diff, err := Diff(source, target)
+		duration := time.Since(start)
+		
+		if err == nil {
+			// Should be fast
+			require.Equal(t, true, duration < time.Second, "Diff creation too slow: %v", duration)
+			
+			// Apply should also be fast
+			start = time.Now()
+			_, err = ApplyDiff(context.Background(), source, diff)
+			duration = time.Since(start)
+			
+			if err == nil {
+				require.Equal(t, true, duration < time.Second, "Diff application too slow: %v", duration)
+			}
+		}
+	})
+}
+
+// PropertyTestDiffSize verifies that diffs are smaller than full states for typical cases
+func FuzzPropertyDiffEfficiency(f *testing.F) {
+	f.Fuzz(func(t *testing.T, slotDelta uint64, numChanges uint8) {
+		if slotDelta > 100 {
+			slotDelta = slotDelta % 100
+		}
+		if numChanges > 10 {
+			numChanges = numChanges % 10
+		}
+		
+		// Create states with small differences
+		source, _ := util.DeterministicGenesisStateElectra(t, 64)
+		target := source.Copy()
+		
+		_ = target.SetSlot(source.Slot() + primitives.Slot(slotDelta))
+		
+		// Make a few small changes
+		if numChanges > 0 {
+			validators := target.Validators()
+			for i := uint8(0); i < numChanges && int(i) < len(validators); i++ {
+				validators[i].EffectiveBalance += 1000
+			}
+			_ = target.SetValidators(validators)
+		}
+		
+		// Create diff
+		diff, err := Diff(source, target)
+		if err != nil {
+			return
+		}
+		
+		// For small changes, diff should be much smaller than full state
+		sourceSSZ, err := source.MarshalSSZ()
+		if err != nil {
+			return
+		}
+		
+		diffSize := len(diff.StateDiff) + len(diff.ValidatorDiffs) + len(diff.BalancesDiff)
+		
+		// Diff should be smaller than full state for small changes
+		if numChanges <= 5 && slotDelta <= 10 {
+			require.Equal(t, true, diffSize < len(sourceSSZ)/2, 
+				"Diff size %d should be less than half of state size %d", diffSize, len(sourceSSZ))
+		}
+	})
+}
+
+// PropertyTestBalanceConservation verifies that balance operations don't create/destroy value unexpectedly
+func FuzzPropertyBalanceConservation(f *testing.F) {
+	f.Fuzz(func(t *testing.T, balanceData []byte) {
+		// Convert byte data to balance changes
+		var balanceChanges []int64
+		for i := 0; i+7 < len(balanceData) && len(balanceChanges) < 50; i += 8 {
+			change := int64(binary.LittleEndian.Uint64(balanceData[i : i+8]))
+			balanceChanges = append(balanceChanges, change)
+		}
+		
+		source, _ := util.DeterministicGenesisStateElectra(t, uint64(len(balanceChanges)+10))
```

Comment:
Can you explain 50 and 10?

Reply:
these are arbitrary just to have a bounded set of data, some is unchanged in the end.